### ISSUE TITLE
Docker crashes when renderer process eats up too much memory

### ISSUE THREAD

**Issue by brian-mann** [cypress-io/cypress#350]
Related to #348 and #349.

When running headlessly on **very long** and memory intense applications we are seeing renderer crashes with Docker.

**Comment by brian-mann** [comment]
This is actually *not* indicative of a memory leak inside of Cypress (nor your application). This has to do with the way that browsers work under the hood.

Luckily, we have a found a simple one line fix for this problem - simply add the flag `--ipc=host`.

[This option is documented here](https://docs.docker.com/engine/reference/run/#/ipc-settings---ipc).

The `Slack app` and `Atom` has also been documented to crash [here](https://github.com/jessfraz/dockerfiles/issues/184, [here](https://github.com/jessfraz/dockerfiles/pull/211), and [here](https://github.com/atom/atom/issues/8684) for the exact same reasons.

If you are using `docker run`

```bash
docker run --ipc=host
``` 

If you are using `docker compose`

```yml
version: '2'
services:
  foobarbazservice:
    ipc: host ## this line right here
```

In the future we are working on a more permanent fix such as described in #349 - either to automatically recover from these crashes, or mostly prevent them up from by nuking the renderer process in between spec files.

**Comment by jheijkoop** [comment]
Hey I seem to be getting this problem, because Chrome is running out of shared memory (`/dev/shm`). By default docker starts images with a 64M `/dev/shm` (try running `df -h` in you instance). To change this you can supply docker with an extra argument: `docker run --shm-size 512M my-image`. Because we are working with mesos/marathon I had to do this via the `"docker": { "parameters": [{"key": "shm-size", "value": "512M"}], ...}` (https://mesosphere.github.io/marathon/docs/native-docker.html) in the json configuration when creating the app/instances.

**Comment by einomi** [comment]
I had the same issue with Codeship CI. Just needed to change config in my codeship-services.yml file, according to documentation https://documentation.codeship.com/pro/continuous-integration/browser-testing/#chrome-crashing

**Comment by egucciar** [comment]
any advice on how to solve for this in CircleCI? I do not have a `services` section in my config.yml, only a `jobs`.

**Comment by mitar** [comment]
I think that since this issue has been made there is now a better fix for the problem by asking Chrome not to use `/dev/shm`. I opened #3633 for more details about this.

**Comment by wralitera** [comment]
hello @brian-mann . How do you set up the --ipc=host into the gitlab-ci.yml?

**Comment by ccorcos** [comment]
Anyone know how to add `--ipc=host` to CircleCI? It looks like they call `docker run` somewhere out of our control...

**Comment by mitar** [comment]
You cannot. And this is why this workaoround does not really work.

**Comment by ccorcos** [comment]
@maximilianschmitt seems to think you can: https://gitter.im/cypress-io/cypress/archives/2018/10/17

**Comment by ccorcos** [comment]
Actually, a build just finished and it crashed pretty quickly

**Comment by ccorcos** [comment]
Looks like there's plenty of shared memory.

```
df -h /dev/shm
Filesystem      Size  Used Avail Use% Mounted on
shm              30G  8.0K   30G   1% /dev/shm
```

**Comment by alk831** [comment]
I had the same issue yet I have changed my image from `cypress/base` to `cypress/browsers:node11.13.0-chrome73` and it now works without crashes on gitlab-ci.
```yml
image: cypress/browsers:node11.13.0-chrome73
```

**Comment by bentrole** [comment]
I will test this image: cypress/browsers:node11.13.0-chrome73 right now. Thanks @alk831

**Comment by bentrole** [comment]
I tested and it failed:
`FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/root/.cache/Cypress/3.2.0/Cypress/libnode.so]
 2: 0x7fa824226887 [/root/.cache/Cypress/3.2.0/Cypress/libnode.so]
 3: 0x7fa823d95a57 [/root/.cache/Cypress/3.2.0/Cypress/libnode.so]
 4: 0x7fa823d959d5 [/root/.cache/Cypress/3.2.0/Cypress/libnode.so]
 5: v8::internal::Factory::NewStruct(v8::internal::InstanceType) [/root/.cache/Cypress/3.2.0/Cypress/libnode.so]
 6: 
<--- Last few GCs --->

[425:0x2659cbff8000]  1853805 ms: Mark-sweep 2057.7 (2176.7) -> 2057.7 (2154.2) MB, 2096.2 / 0.0 ms  (+ 0.0 ms in 0 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 2096 ms) last resort 
[425:0x2659cbff8000]  1856036 ms: Mark-sweep 2057.7 (2154.2) -> 2057.7 (2154.2) MB, 2230.4 / 0.0 ms  last resort 


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x162eb08ad681 <JSObject>
    1: bindContainer(aka bindContainer) [/builds/wynd-qa/BO/node_modules/typescript/lib/typescript.js:~22229] [pc=0x2ef0ae593de3](this=0x2ae892502311 <undefined>,node=0x7235b368b9 <NodeObject map = 0x3516afdfd7d1>,containerFlags=45)
    2: bind(aka bind) [/builds/wynd-qa/BO/node_modules/typescript/lib/typescript.js:~23556] [pc=0x2ef0ae55cdac](this=0x2ae892502311 <undefined>,node=0x7235b368b9 <No...

v8::internal::Factory::NewTuple3(v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>) [/root/.cache/Cypress/3.2.0/Cypress/libnode.so]
 7: 0x7fa823b407a1 [/root/.cache/Cypress/3.2.0/Cypress/libnode.so]
 8: 0x7fa823b4347b [/root/.cache/Cypress/3.2.0/Cypress/libnode.so]
 9: 0x7fa823b431bc [/root/.cache/Cypress/3.2.0/Cypress/libnode.so]
10: 0x7fa823b47b5f [/root/.cache/Cypress/3.2.0/Cypress/libnode.so]
11: 0x2ef0add043fd

Timed out waiting for the browser to connect. Retrying...

The following error was thrown by a plugin. We've stopped running your tests because a plugin crashed.`

**Comment by alk831** [comment]
@bentrole It crashed only once for me with this image and now it works properly. I would try to split large tests into smaller ones and run without recording until there is no fix for this.

**Comment by bentrole** [comment]
@alk831 I already splited my tests because of that but this bug is there since 2016, If there is no fixes until now, I think that they will not correct that! ;-(

**Comment by ccorcos** [comment]
@brian-mann can we please reopen this issue? If we cannot scale out our testing library, I don't think we can use this service :/

**Comment by alk831** [comment]
@bentrole one more thing that you can try is to run test in parallel so memory usage is smaller, however
as @brian-mann mentioned this is not strictly related with cypress and maybe your CI runner is just not capable of running chromium. If you are using gitlab CI then you could try to create your own runner.

**Comment by wralitera** [comment]
My tests are structured like that: 1 spec file = 1 scenario. So when I run all tests using the --spec option, each spec file should run one by one and should not make an out of memory.

Running in parallele works but I have not my own runner and if I use too many runner and the test take time, I ll block other developers.

I will ask to our admin system (of gitlab runner) to try to custom the runner and activate the --ipc=host suggested by @brian-mann

**Comment by Ameobea** [comment]
Bumping to `cypress/browsers:node11.13.0-chrome73` as my base Docker image fixed this for me as well!  🎊

**Comment by r3dDoX** [comment]
For people struggling to set ipc=host on their respective environment it should help to mount a volume to /dev/shm to allow shared memory for POSIX objects.
In Kubernetes/OpenShift for example I mounted an "EmptyDir" volume to /dev/shm and everything is working smoothly now.

**Comment by jbinto** [comment]
This seems to have gotten way worse in 3.3.0. We've been running Cypress on CircleCI for almost 2 years, and these crashes have become way more frequent in the last few weeks.

**Comment by jbinto** [comment]
We're seeing Cypress crashes multiple times daily now in CircleCI after a couple of years of solid reliability. None of the workarounds here have helped:

* Cannot set `--ipc=host` on Circle
* Using `cypress/browsers:node11.13.0-chrome73` also had crashes
* `--disable-dev-shm-usage` is not only available since Chrome 65, and Electron is only 61

We've *never* seen a "sad face" crash since about 1 month ago. I'll look back at our CI logs and see when it started.

**Comment by jbinto** [comment]
It looks like crashes started 8 days ago since we upgraded to `3.3.1`. I don't see any crashes associated with `3.3.0`. This could still all be coincidence but for now we'll try downgrading the Cypress package.

**Comment by idanRiski** [comment]
Any news with this issue? 
I've already tried all your suggestions to use the flages of "ipc=host", "cpus=2"  or to increase the shm/dev memory.. also to downgrade the cypress to 3.2.0 and to add the flages below to the cypress.json file:   "numTestsKeptInMemory": 0, "restartBrowserBetweenSpecFiles": true, "videos: false".. but none of them has solved this issue.
I got this error message (crashed becasue a sad face.." in every execution in time it using "cy.visit".
this issue happens only by docker and didn't recreat by locally execution .

please your assistance, thanks.

**Comment by jbinto** [comment]
We have switched to `--browser chrome` and we disable /dev/shm usage as so:

```js
// cypress/plugins/index.js

module.exports = (on, config) => {
  // ref: https://docs.cypress.io/api/plugins/browser-launch-api.html#Usage
  on('before:browser:launch', (browser = {}, args) => {
    if (browser.name === 'chrome') {
      args.push('--disable-dev-shm-usage')
      return args
    }

    return args
  })
}

```

And crashes have stopped completely. I suspect something changed in between Electron 59 and Electron 61 that makes it more prone to crashes, but for anyone hitting this on CircleCI this is the only actual fix (after weeks of tinkering).

**Comment by Ameobea** [comment]
@jbinto this looks really promising; I don't have access to the K8s cluster on which our build node runs, so I can't tweak anything on that end.  I'll give this a shot and see if it solves our crashes as well!

**Comment by amkoehler** [comment]
@jbinto I had about perfect timing with opening this issue and seeing your solution. This is working well for us so far using the `circleci/node:8-browsers` image

**Comment by yuanyuanlimaggie** [comment]
Anyone has a good solution for Gitlab? I updated to `cypress/browsers:node11.13.0-chrome73` in Gitlab. It still crashed. I am using Cypress 3.3.2.

**Comment by mitar** [comment]
@jbinto But without chrome you do not get recordings.

**Comment by jennifer-shehane** [comment]
Going to reopen this since people are still having issues with crashes in Docker. 

Our plan of action now is to investigate passing in the `--disable-dev-shm-usage` automatically. Related to https://github.com/cypress-io/cypress/issues/3633

## Workaround Today

Follow these instructions for passing the `--disable-dev-shm-usage` flag: https://github.com/cypress-io/cypress/issues/350#issuecomment-503231128

cc @flotwig

**Comment by srinu-kodi** [comment]
Hi @brian-mann @jennifer-shehane 
Adding the --ipc=host for docker run is also not helping out and specs are failing continuously.

We have integrated the tests in CI and specs are failing. Is there any solution in near future as it is giving hard times now...

Thanks

**Comment by you1anna** [comment]
This is now blocking us using Cypress in CI, have tried adding `--disable-dev-shm-usage` as @jennifer-shehane suggested but we're seeing the 'Chromium renderer process crashed' message each time in GoCD, after Cypress has run only two tests.

**Comment by srinu-alef** [comment]
> This is now blocking us using Cypress in CI, have tried adding `--disable-dev-shm-usage` as @jennifer-shehane suggested but we're seeing the 'Chromium renderer process crashed' message each time in GoCD, after Cypress has run only two tests.

Hi @you1anna I also faced this problem, and today Finally I could fix this by adding 

> shm_size: '2gb'

 in my docker-compose.yaml file and it is working consistently.
Basically this flag will allocate the 2gb memory to the docker container and such that there will be no memory related issue...

Please add this and it should be fine. BTW it doesn't have any link with GoCD :)

Let me know if you face any problem or need help.

**Comment by srinu-kodi** [comment]
> This is now blocking us using Cypress in CI, have tried adding `--disable-dev-shm-usage` as @jennifer-shehane suggested but we're seeing the 'Chromium renderer process crashed' message each time in GoCD, after Cypress has run only two tests.

Hi @you1anna I also faced this problem, today finally I could fix this by adding `shm_size: '2gb'`
in my docker-compose.yaml file and it is working consistently.
Basically this flag will allocate the 2gb memory to the docker container and such that there will be no memory related issue...

Please add this and it should be fine. BTW it doesn't have any link with GoCD :)

Let me know if you face any problem or need help.

**Comment by you1anna** [comment]
Thanks @srinu-kodi. It makes sense to use more shm, but in our tests we are using a separate container with following volume share:
   ` volumes:
        - "/dev/shm:/dev/shm"`
which means that all of shared memory is used by the container. Currently we are using agents with 8GB ram, so our next step is to try with 16GB. Outside of that I don't know what else we can do.

**Comment by mvandebunt** [comment]
have this problem since using 3.4.1
running dockers in a wercker pipeline

**Comment by cmcnicholas** [comment]
recently upgraded to 3.4.1 and we are also exhibiting the same issues here pretty much every test run now (because we have parallel tests) results in 2-3 machines failing whilst the other 4 pass.

**Comment by fulvio-m** [comment]
For Gitlab users the fix requires editing of the runner configuration by adding a new volume to the `volumes` configuration of your docker runner:
```
[[runners]]
  ...
  executor = "docker"
  ...
  [runners.docker]
   ...
    volumes = ["/dev/shm:/dev/shm", ...other volumes...]
  ...
  [runners.cache]
```

Basically this instructs the docker instance to mount the underlying `/dev/shm` volume.
After editing runner needs to be restarted with `gitlab-runner restart`.
Credits for the fix to https://lcx.wien/blog/cypress-gitlab-ci/

**Comment by mohansgithub** [comment]
Any idea how to use --ipc==host on the dynamic docker based jenkins agents ( which will be deployed on Kubernetes) ? in below configuration we specify the cypress image and we need to use --ipc==host

![image](https://user-images.githubusercontent.com/25417816/68356127-81c19d00-0137-11ea-9e40-93681ec0827b.png)

**Comment by Bsmalhi** [comment]
@mohansgithub use Environment Variables add "ipc" as key and "host" as value that will do it for your docker container. Let me know if you figured it out.

**Comment by karthikkota** [comment]
Thank you @fulvio-m.

**Comment by JimmyKuruvilla** [comment]
`--ipc=host` works locally for me, but as an env var in the jenkins kubernetes plugin it did not work. However mounting it as a volume did work:
![image](https://user-images.githubusercontent.com/7905473/70083208-e28fa880-1603-11ea-94df-eec4497e61e3.png)

**Comment by hannahhaken** [comment]
We're running Cypress 3.8.1 in Electron 78 in Docker across multiple apps and have been seeing this issue for a couple of weeks in one particular app. The rest of the apps were fine. We tried the suggested IPC host workaround but this didn't help....

What _has_ fixed this issue is removing multiple uses of `describe`. So it appears this particular test suite had multiple instances of `describe()` nested inside a single `describe()`. I've now replaced these nested `describe()` with `it()`. The memory leak issue has subsided and this test suite now finishes without any problems.

**Comment by 9odzilla** [comment]
@hannahhaken My tests are already written with top level `describe()` and `it()` within them. I still see this issue using electron.

**Comment by bigbitbus** [comment]
We have the same issue with Google Cloudbuild - its not possible to specify `ipc=host` within the Google Cloudbuild specification file.

**Comment by zeel-swiggy** [comment]
This is not happening for 3.6.0, but happened for 3.7.0 and 3.8.0

**Comment by must-git-good** [comment]
Joining the various voices that are pointing out that this error seems to be cropping up for us for the first time over the last few weeks.  Didn't happen on all tests, but did consistently happen on some (that shouldn't have actual memory issues).  Losing nested describe blocks helped, but errors remained (all this despite running ipc=host in Docker, etc.). It seems like swapping to Chrome for now may help, which means Electron and Cypress may have some interesting optimization/interactions going on under the hood.

No other input, will update if I find out more, but want to make sure this gets looked at, as it's been a significant frustration for us.

**Comment by lsdkjflasjflkdasj** [comment]
Same thing happening here.

**Comment by tommueller** [comment]
We are also constantly running into this issue and by now this is creating a lot of frustration here.
We are running the Cypress tests in a Drone-CI docker setup. We already tried these recommendations:
- increased the size of dev/shm
- upgraded to latest versions of drone, drone-agent and cypress
- decreased numTestsKeptInMemory setting
without any success.
I have not yet tried these options, as they are not possible with our current setup or seem to be too unstable ...
- add --ipc=host on the docker container, which is currently not possible because drone does not support it for now, there is also an open ticket for this: https://discourse.drone.io/t/does-drone-support-ipc-host-option/1049
- try using the --disable-dev-shm-usage-flag on Chrome https://github.com/cypress-io/cypress/issues/5336

Any updates on this would be a great help for us!

**Comment by davidzambrana** [comment]
We fixed this problem after implementing a few recommendations we saw here:

- Updated to Cypress 3.8.0
- Moved to this image `cypress/browsers:node10.16.0-chrome77`
- Started using `-headless -b chrome` in the run command
- Updated our `plugins/index` file with
```
on('before:browser:launch', (browser = {}, args) => {
    if (browser.name === 'chrome') {
      args.push('--disable-dev-shm-usage')
      return args
    }

    return args
  })
```

The crashes stopped in all the test collections we have (~15)

**Comment by tommueller** [comment]
I can confirm that the changes @davidzambrana proposed also worked for us so far. 
We now run into the issue described here https://github.com/cypress-io/cypress/issues/1297 but this is probably more an issue of how we wrote the tests.
I will try to fix the tests and keep an eye on this. Thanks @davidzambrana for now!

**Comment by mulholio** [comment]
To anyone wondering what the `--disable-dev-shm-usage` flag does, this page is a useful explainer https://developers.google.com/web/tools/puppeteer/troubleshooting#tips

**Comment by shlima** [comment]
https://github.com/cypress-io/cypress/issues/350#issuecomment-574072211 can confirm the solution, but I've changed `cypress/browsers:node10.16.0-chrome77` image to `cypress/included:4.0.2` as it currently includes `cypress` and `chrome`.

My specs started to run smoothly on Gitlab CI with cypress in docker-in-docker

**Comment by adinizs** [comment]
We resolved this problem using the image `cypress/browsers:node10.16.0-chrome` on dockerfile and running chrome headless.

**Comment by radDunin** [comment]
Hi,
i can also confirmed @davidzambrana  solution, what is important is that the image should be this one: cypress/browsers:node10.16.0-chrome77 it works even in headed chrome. For now thats doesn't works on chrome 80 and chrome 69

**Comment by otalandim** [comment]
> I had the same issue yet I have changed my image from `cypress/base` to `cypress/browsers:node11.13.0-chrome73` and it now works without crashes on gitlab-ci.
> 
> ```yaml
> image: cypress/browsers:node11.13.0-chrome73
> ```

For me is not works

**Comment by otalandim** [comment]
> We fixed this problem after implementing a few recommendations we saw here:
> 
> * Updated to Cypress 3.8.0
> * Moved to this image `cypress/browsers:node10.16.0-chrome77`
> * Started using `-headless -b chrome` in the run command
> * Updated our `plugins/index` file with
> 
> ```
> on('before:browser:launch', (browser = {}, args) => {
>     if (browser.name === 'chrome') {
>       args.push('--disable-dev-shm-usage')
>       return args
>     }
> 
>     return args
>   })
> ```
> 
> The crashes stopped in all the test collections we have (~15)

For me it doesnot works

Another problem happened...can you help me please

>>>>
  CypressError: cy.visit() failed trying to load:
 https://www.americanas.com.br/
 We attempted to make an http request to this URL but the request failed without a response.
 We received this error at the network level:
   > Error: ESOCKETTIMEDOUT
 Common situations why this would fail:
   - you don't have internet access
   - you forgot to run / boot your web server
   - your web server isn't accessible
   - you have weird network configuration settings on your computer
 The stack trace for this error is:
 Error: ESOCKETTIMEDOUT
     at ClientRequest.<anonymous> (/root/.cache/Cypress/4.2.0/Cypress/resources/app/packages/server/node_modules/request/request.js:816:19)
     at Object.onceWrapper (events.js:299:28)
     at ClientRequest.emit (events.js:210:5)
     at TLSSocket.emitRequestTimeout (_http_client.js:690:9)
     at Object.onceWrapper (events.js:299:28)
     at TLSSocket.emit (events.js:210:5)
     at TLSSocket.Socket._onTimeout (net.js:468:8)
     at listOnTimeout (internal/timers.js:531:17)
     at processTimers (internal/timers.js:475:7)
 Because this error occurred during a 'before each' hook we are skipping the remaining tests in the current suite:

**Comment by przemuh** [comment]
Switching to Chrome with `--disable-dev-shm-usage` as @davidzambrana suggested - works for me!

Thanks a lot! 🎉 

@otalandim Please remember that since Cypress 4.0.0 there is a change in passing arguments to the browser (https://docs.cypress.io/guides/references/migration-guide.html#Plugin-Event-before-browser-launch)

I use GitlabCI. Here are my changes:

GitlabCI config
```
-  image: cypress/base:12.16.2
+  image: cypress/browsers:node12.16.2-chrome81-ff75
````

cypress/plugins/index.js
```
on("before:browser:launch", (browser = {}, launchOptions) => {
            if (browser.name === "chrome") {
                launchOptions.args.push("--disable-dev-shm-usage");
                return launchOptions;
            }
}
```

run command:
```
cypress run --browser chrome --headless
```

**Comment by ttomaszewski** [comment]
Any idea when this will get resolved for Electron?

**Comment by digitalkaoz** [comment]
for anyone getting here and want to know how to fix it on gitlab-ci kubernetes executor:

```
$ kubectl edit cm gitlab-runner-gitlab-runner
```

config part to add right before `# Start the runner` in entrypoint
see https://gitlab.com/gitlab-org/gitlab-runner/-/issues/2578#note_186055981

```
    cat >> /home/gitlab-runner/.gitlab-runner/config.toml << EOF
          [[runners.kubernetes.volumes.empty_dir]]
              name = "shm"
              mount_path = "/dev/shm"
              medium = "Memory"
    EOF
```

**Comment by YeeJone** [comment]
> Switching to Chrome with `--disable-dev-shm-usage` as @davidzambrana suggested - works for me!
> 
> Thanks a lot! 🎉
> 
> @otalandim Please remember that since Cypress 4.0.0 there is a change in passing arguments to the browser (https://docs.cypress.io/guides/references/migration-guide.html#Plugin-Event-before-browser-launch)
> 
> I use GitlabCI. Here are my changes:
> 
> GitlabCI config
> 
> ```
> -  image: cypress/base:12.16.2
> +  image: cypress/browsers:node12.16.2-chrome81-ff75
> ```
> 
> cypress/plugins/index.js
> 
> ```js
> on("before:browser:launch", (browser = {}, launchOptions) => {
>   if (browser.name === "chrome") {
>     launchOptions.args.push("--disable-dev-shm-usage");
>     return launchOptions;
>   }
> }
> ```
> 
> run command:
> 
> ```
> cypress run --browser chrome --headless
> ```

I use ` cypress.run() `  api to run cypress,  so how do I add  the `--headless`

**Comment by sabarishnarain** [comment]
This is still a problem in Azure CI with electron (cypress 4.11) when using the base docker images. The workaround is to use chrome by following @przemuh instructions.

**Comment by keul** [comment]
For posterity: I've this same issue in Gitlab CI.

Unluckily every `gitlab-ci.yml` workaround above require the change of the base docker image, while I'm not (and I can't) using any of them: I am **manually installing cypress on a Centos** (is an internal gitlab installation using a private runner server).

In my case this [gitlab comment to issue](https://gitlab.com/gitlab-org/gitlab-runner/-/issues/2168#note_190617588) solved the problem.

**Comment by CoryDanielson** [comment]
There's a chrome flag that exposes `window.gc()` which forces garbage collection.

```console
--js-flags=--expose-gc
```
### plugins/index.js
```js
module.exports = (on, config) => {
	on('before:browser:launch', (browser, launchOptions) => {
		if (browser.name === 'chrome') {
			// exposes window.gc() function that will manually force garbage collection
			launchOptions.args.push('--js-flags=--expose-gc');
		}

		return launchOptions;
	});
};
```
### support/index.js
```js
afterEach(() => {
	cy.window().then(win => {
		// window.gc is enabled with --js-flags=--expose-gc chrome flag
		if (typeof win.gc === 'function') {
			// run gc multiple times in an attempt to force a major GC between tests
			win.gc();
			win.gc();
			win.gc();
			win.gc();
		}
	});
});
```

### Anecdotal Data - max memory in MB during test run
| win.gc once | win.gc 4x | no win.gc |
| --- | --- | --- |
| 299 | 267 | 535 |
| 406 | 369 | 585 |
| 372 | 278 | 598 |
| 445 | 298 | 570 |
| --- | --- | --- |
| 380 | 303 | 572 |

303/572 = ~53%

From my anecdotal evidence running a big test against a big memory intensive app, using `window.gc()` reduced memory consumption by roughly 50% during `cypress open` (with `numTestsKeptInMemory: 0`) I suspect that this will reduce crashes as well when running in CI.

I also found that running `window.gc()` multiple times in a row seemed to be much more effective than running it just once. I think this causes chrome to trigger a major GC event rather than a minor one.

I saw several GCs that reduced memory by 100-200mb which was pretty cool.

**Comment by phillyc** [comment]
For those of you who are trying to solve this issue in **GitLab,** on the free tier, I found a simple solution.

I was trying to create some smoke tests for my team's application, and noticed that the GitLab CICD runner would typically get through the first 2-3 tests of each test file before crashing with the above out of memory symptoms.

**Condense all of the actions and test assertions into 1-2 it('can do foo') test statements.** 

There appears to be a pretty big memory overhead in the setup/teardown of each of these cases. I was able to take around 60 test cases in my locally passing integration test stack and condense them down into 10 test.js files with 1 test case each. 

In summary, **try putting all your tests into one test case**.

**Comment by jackbravo** [comment]
What about when running on github actions, like https://github.com/cypress-io/github-action ? How do you enable `ipc=host`?

**Comment by mattvb91** [comment]
Started experiencing this again on github actions upgrading from 5.3 to 6.2

**Comment by vincentpalita** [comment]
Version 6.2.1 seems to fix that for me.

**Comment by mattvb91** [comment]
Unfortunately 6.2.1 still crashing https://github.com/corejam/corejam/runs/1650892998?check_suite_focus=true#step:9:290 staying on 5.3 for now. Maybe someone else can chime in a version number when it started happening for them between 5.3 & 6.2 as I dont have the time to manually go through each version / build the associated docker images

**Comment by mattvb91** [comment]
https://github.com/corejam/corejam/runs/1784204472?check_suite_focus=true#step:9:313

Still no luck on 6.3.0

5.3.0 still stable for us have not had any chrome crashes there

**Comment by mattvb91** [comment]
Anyone looking to enable `IPC=host` on github actions it looks like it is possible like this: 
https://github.com/corejam/corejam/pull/311/files#diff-014228303dff9a1af15f4bbd18401f906380129b10ae2a2c62f8b8be592ff88eR105

You can see it setting the option here: https://github.com/corejam/corejam/runs/1888917637?check_suite_focus=true#step:2:111 
` /usr/bin/docker create --name 01f73992fe8148a89514be690697b85d_corejamcypress640_aa44b9 --label 5588e4 --workdir /__w/corejam/corejam --network github_network_3bce8c6715944b47a995b42f42973efa --ipc=host    ...`

More context here: https://github.com/actions/virtual-environments/discussions/2697#discussioncomment-362677

Just let it run through 3 times on our setup and its finally running it seems!! 

Can anyone else confirm? If its working for others ill put in a MR with the above sample for docs

**Comment by dobriai** [comment]
FWIW, in our case adding `--shm-size=512m` seems to have quieted down the issue. For now. Using cypress@6.4.0.

**Comment by TimothyRHuertas** [comment]
If you added --ipc=host and you still get the error, try restarting docker 🤦

**Comment by flotwig** [comment]
In 7.0.1, Cypress passes `disable-dev-shm-usage` to Chrome and Electron on startup, which should remove the need to use `--ipc=host` or increase the `/dev/shm` size. Once it's released this should be solved.
